### PR TITLE
Navi Carbon Querylogger

### PR DESCRIPTION
**Pull request by dvarshney1**
## Description
Added QueryLogging Implementation for Carbon in Navi.

## Proposed Changes
In package/webservice, the following files were added to reflect this change:

- CarbonQueryLogger.kt: Has the implementation for Query Logging and overrides the QueryLogger interface from Elide.
- QueryStatus.kt: Enum class with all the possible query statuses that a query could have when an response is received from Elide
- QueryStats.kt: Our querylogging entity class.

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        user: User?,
+        headers: MutableMap<String, String>?,
+        apiVer: String?,
+        queryParams: Optional<MultivaluedMap<String, String>>?,
+        path: String?
+    ) {
+        var start: Long = System.currentTimeMillis()
+        var apiQuery: String? = constructAPIQuery(queryParams, path)
+        var userName: String? = user?.name
+        synchronized(queryWatcher) {
```
Comment (aklish) [inline packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:74]: You can use a ConcurrentHashMap and removed all the synchronized blocks that protect queryWatcher.

File: packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        var qstat: QueryStats = queryWatcher[queryId]!!
+        synchronized(this) {
+            qstat.status = QueryStatus.CANCELLED
+            qstat.durationInMillis = endTime - qstat.startTime!!
+            qstat.bytesReturned = 0
+            qstat.rowsReturned = 0
+            queriesToBeLogged.add(qstat)
+            queryWatcher.remove(queryId)
+        }
+        if (queriesToBeLogged.size > maxQueries) { run() }
```
Comment (aklish) [inline packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:60]: This should be moved inside the synchronized block.

File: packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        var status: QueryStatus = if (response.data == null) QueryStatus.FAILED else QueryStatus.COMPLETED
+        synchronized(this) {
+            qstat.status = status
+            qstat.durationInMillis = endTime - qstat.startTime!!
+            qstat.bytesReturned = getTotalBytesReturned(response)
+            qstat.rowsReturned = getTotalRowsReturned(response)
+            queriesToBeLogged.add(qstat)
+            queryWatcher.remove(queryId)
+        }
+        if (queriesToBeLogged.size > maxQueries) { run() }
```
Comment (aklish) [inline packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:98]: Should be moved inside the synchronized block

File: packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            qstat.modelName = query.table.name
+            qstat.storeQuery = apiQuery
+            qstat.status = QueryStatus.INPROGRESS
+            qstat.isCached = isCached
+            queryWatcher[queryId] = qstat
+        }
+    }
+
+    override fun run() {
+        synchronized(queriesToBeLogged) {
```
Comment (aklish) [inline packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:113]: Should be synchronized(this)

File: packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            for (qs in queriesToBeLogged) {
+                entityManager.persist(qs)
+            }
+            queriesToBeLogged.clear()
+        }
+    }
+
+    private fun constructAPIQuery(queryParams: Optional<MultivaluedMap<String, String>>?, path: String?): String? {
+        var apiQuery: String?
+        if (queryParams == null) {
```
Comment (aklish) [inline packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:123]: Just curious - when is this null?

Comment (dvarshney1) [reply packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:123]: Updated. This is never null

**Comment by aklish** [comment]
Hey @dvarshney1 .  Do you plan on completing this PR or should we assign someone to wrap it up?

**Comment by dvarshney1** [comment]
Hi Aaron, I plan on finishing this. Got a bit busy with school work but I will work on it during this weekend. I will let you know when this is done. @aklish

File: packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/utilities/QueryStats.kt [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+import java.util.Date
+import java.util.UUID
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.Id
+//import javax.persistence.Table
+import javax.persistence.Temporal
+import javax.persistence.TemporalType
+
+@Include(rootLevel = true)
```
Comment (aklish) [inline packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/utilities/QueryStats.kt:25]: This should not be an elide entity.  It should be a JPA entity only.

File: packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/utilities/QueryStats.kt [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+@Entity
+@Cardinality(size = CardinalitySize.LARGE)
+@VersionQuery(sql = "SELECT COUNT(*) from QueryStats")
```
Comment (aklish) [inline packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/utilities/QueryStats.kt:28]: Remove all the Elide annotations from this model.

**Review by aklish — CHANGES_REQUESTED** [review]

File: packages/webservice/app/build.gradle.kts [hunk 8]
```diff
@@ -16,10 +16,11 @@ repositories {
 dependencies {
     implementation(project(":models"))
     implementation("org.springframework.boot:spring-boot-starter-security")
-    implementation("com.yahoo.elide", "elide-spring-boot-starter", "5.0.0-pr12")
+    implementation("com.yahoo.elide", "elide-spring-boot-starter", "5.0.0-pr15")
```
Comment (aklish) [inline packages/webservice/app/build.gradle.kts:19]: we are on pr21 now.  This PR will need to be rebased on master.

File: packages/webservice/app/src/main/resources/models/tables/queryStats.hjson [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            '''
+            Carbon Query Stats Table
+            '''
+            cardinality: large
+            measures :
+            [
+                {
+                    name: durationInMillis
+                    type: DECIMAL
+                    definition: IF (durationInMillis IS NULL, -1, durationInMillis)
```
Comment (aklish) [inline packages/webservice/app/src/main/resources/models/tables/queryStats.hjson:17]: Check out the latest documentation for HJSON configuration:
https://elide.io/pages/guide/v5/04-analytics.html

Any column referenced in the definition should be wrapped in {{ and }}.  So definition here should be:
`IF ({{durationInMillis}} IS NULL, -1, {{durationInMillis}})`

File: packages/webservice/app/src/main/resources/models/tables/queryStats.hjson [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+                    type: DECIMAL
+                    definition: IF (bytesReturned IS NULL, -1, bytesReturned)
+                }
+            ]
+            dimensions :
+            [
+                {
+                    name: requestId
+                    type: TEXT
+                    definition: requestId
```
Comment (aklish) [inline packages/webservice/app/src/main/resources/models/tables/queryStats.hjson:35]: Same definition comment for dimension definitions (wrap column names in {{ and }}).

File: packages/webservice/app/src/main/resources/models/tables/queryStats.hjson [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            description:
+            '''
+            Carbon Query Stats Table
+            '''
+            cardinality: large
+            measures :
+            [
+                {
+                    name: durationInMillis
+                    type: DECIMAL
```
Comment (aklish) [inline packages/webservice/app/src/main/resources/models/tables/queryStats.hjson:16]: Type should be a INTEGER

File: packages/webservice/app/src/main/resources/models/tables/queryStats.hjson [hunk 12]
```diff
@@ -0,0 +1,113 @@
+{
+    tables:
+    [
+        {
+            name: QueryStats
+            table: QueryStats
+            description:
+            '''
+            Carbon Query Stats Table
```
Comment (aklish) [inline packages/webservice/app/src/main/resources/models/tables/queryStats.hjson:9]: Change Carbon to Navi

File: packages/webservice/app/src/main/resources/models/tables/queryStats.hjson [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+                    type : TEXT
+                    definition : IF (hostName IS NULL, 'Unknown', hostName)
+                }
+                {
+                    name : createdOn
+                    type : TIME
+                    definition : createdOn
+                    grains: [
+                    {
+                        grain :  MONTH
```
Comment (aklish) [inline packages/webservice/app/src/main/resources/models/tables/queryStats.hjson:68]: Elide currently defines the following time grains:

    SIMPLEDATE("yyyy-MM-dd"),
    DATETIME("yyyy-MM-dd HH:mm:ss"),
    MONTHYEAR("MMM yyyy"),
    YEARMONTH("yyyy-MM"),
    YEAR("yyyy"),
    WEEKDATEISO("yyyy-MM-dd")

File: packages/webservice/models/build.gradle.kts [hunk 14]
```diff
@@ -29,9 +29,11 @@ allOpen {
 }
 
 dependencies {
-    implementation("com.yahoo.elide", "elide-core", "5.0.0-pr12")
+    implementation("com.yahoo.elide", "elide-core", "5.0.0-pr15")
+    implementation("com.yahoo.elide", "elide-datastore-aggregation", "5.0.0-pr15")
```
Comment (aklish) [inline packages/webservice/models/build.gradle.kts:33]: See earlier comment about upgrading elide and rebasing.

File: packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/utilities/QueryStats.kt [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    var createdOn: Date? = null
+
+    @Column
+    var hostName: String? = null
+
+    @Transient
+    var startTime: Long? = null
+}
+
+/*
```
Comment (aklish) [inline packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/utilities/QueryStats.kt:85]: Let's remove the HJSON comment here.

File: packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+import javax.transaction.Transactional
+import javax.ws.rs.core.MultivaluedMap
+
+object Constants {
+    const val MAX_QUERIES = 5
+    const val TIMEOUT = 10000
+}
+
+@Transactional
+open class CarbonQueryLogger constructor(
```
Comment (aklish) [inline packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:31]: Maybe change name to NaviQueryLogger.  Carbon isn't a public name.

File: packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        var status: QueryStatus = if (response.data == null) QueryStatus.FAILED else QueryStatus.COMPLETED
+        synchronized(this) {
+            qstat.status = status
+            qstat.durationInMillis = endTime - qstat.startTime!!
+            qstat.bytesReturned = getTotalBytesReturned(response)
+            qstat.rowsReturned = getTotalRowsReturned(response)
+            queriesToBeLogged.add(qstat)
+            queryWatcher.remove(queryId)
+        }
+        if (queriesToBeLogged.size > maxQueries) { run() }
```
Comment (aklish) [inline packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:98]: Fetch queriesToBeLogged.size() inside the synchronized block.  YOu can use the value outside as you do here.

Comment (aklish) [reply packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt:98]: Alternatively, move the entire condition inside the synchronized block.

**Review by aklish — CHANGES_REQUESTED** [review]

File: packages/webservice/app/src/main/resources/models/tables/queryStats.hjson (hunk 13 above)
Comment (jkusa) [reply packages/webservice/app/src/main/resources/models/tables/queryStats.hjson:68]: Here is the new set of grains: https://github.com/yahoo/elide/pull/1608

**Comment by jkusa** [comment]
Hey @dvarshney1, happy new year! Can you please rebase your PR?

**Comment by jkusa** [comment]
Closing for now

**Pull request closed without merging**